Use VideoPress videos as source of the core video blocks

mmtr opened the pull request:
#### Changes proposed in this Pull Request:
This PR ensures that we are using the VideoPress videos on published posts containing video blocks.

* Add a new `VideoPress_Gutenberg` class to the VideoPress module that defines the server-side rendering of the `core/video` block.
* The server-side rendering returns the block content as defined by the client-side but replacing the `src` attribute used in the video player with the VideoPress URL if it is available.
* The logic has been created in a new file `class.videopress-gutenberg.php` so it can be reused in WP.com and synced with Fusion (D23482-code).

#### Testing instructions:

* Spin up a new Jurassic Ninja test site: https://jurassic.ninja/create/?jetpack-beta&branch=add/videopress-src-core-video-blocks&shortlived&wp-debug-log&gutenpack
* Connect Jetpack to a WP.com site with a premium/business plan.
* Go to Jetpack → Settings and activate the VideoPress module (by enabling the "Enable high-speed, ad-free video player" option).
* Add a new post.
* Insert a video block.
* Upload a video file using the Media Library in a format that is not supported by your browser (e.g. .mov in Chrome). 
  * VideoPress only converts videos uploaded with the Media Library, that's why we need to use that instead of the "Upload" button (to be fixed in a follow-up PR/commit: https://github.com/Automattic/jetpack/issues/11194).
<img width="673" alt="screen shot 2019-01-15 at 17 11 12" src="https://user-images.githubusercontent.com/1233880/51193290-cd753280-18e8-11e9-8417-b69070f3109c.png">


* Publish it immediately.
* Check that a black placeholder is displayed on the published view. Note that this will only happen if VideoPress has not finished yet to convert the video.
<img width="333" alt="screen shot 2019-01-14 at 16 42 41" src="https://user-images.githubusercontent.com/1233880/51122999-6f2e4e00-181b-11e9-9583-05f3cb7583bb.png">

* Wait until VideoPress finishes to convert the video (you can check this on the Media section).
* Reload the published view of the post.
* Make sure that the video is now playable.
<img width="547" alt="screen shot 2019-01-14 at 16 48 22" src="https://user-images.githubusercontent.com/1233880/51123371-3b075d00-181c-11e9-900a-99940599b4ad.png">

* Inspect the video player with the browser devtools and confirm that the `src` attribute is pointing to the VideoPress video (`videos.files.wp.com`).
* Repeat the process downgrading the plan or disabling the VideoPress module and check that the video is not playable on the published view and the `src` attribute doesn't point to `videos.files.wp.com`. _Note that you need to upload a new video since the previous one is already converted)._

#### Proposed changelog entry for your changes:

* Use VideoPress videos as source of the core video blocks

mmtr committed:
Use VideoPress videos as source of the core video blocks

mmtr committed:
Make function static

dmsnell commented on modules/videopress/class.videopress-gutenberg.php:
```diff
@@ -0,0 +1,84 @@
+	public static function init() {
+		if ( self::should_initialize() ) {
+			add_action( 'init', array( __CLASS__, 'register_video_block_with_videopress' ) );
+		}
```
pardon me for not knowing Jetpack patterns as much, but why are we calling `should_initialize()` from inside `init()`? is there a reason not to include the `should_initialize()` logic here in this function and then conditionally load the action if the conditions hold?

mmtr replied:
I only did this in order to follow the same pattern I observed in the `Jetpack_Gutenberg` class: https://github.com/Automattic/jetpack/blob/535f34b30532ea1fe39934f1201ac760b98ca447/class.jetpack-gutenberg.php#L174

dmsnell replied:
let's see if @roccotripaldi has any thoughts here. it looks like this may have been [done in a refactor for the sake of reusing a common condition](https://github.com/Automattic/jetpack/pull/10893) among many places. if we don't have a strong reason for copying that here how about we just remove the layer and keep it straightforward in `init`?

mmtr replied:
After confirming that only the Gutenberg check was the only one really needed (this code is never executed if Jetpack is inactive and the module is always loaded in WP.com if we have the right plan), I simplified everything, removing the `should_initialize` function and moving the Gutenberg check directly to the `init` function.

gwwar commented:
Seeing the following error on this branch, on a paid plan, activated recommended modules, then activated the video module.

<img width="1171" alt="screen shot 2019-01-24 at 1 49 10 pm" src="https://user-images.githubusercontent.com/1270189/51710716-d8ae1980-1fde-11e9-951b-00b679d4bd95.png">
```
Notice: WP_Block_Type_Registry::register was called incorrectly. Block type "jetpack/simple-payments" is already registered. Please see Debugging in WordPress for more information. (This message was added in version 5.0.0.) /public/wp-includes/functions.php on line 4231
```

gwwar commented:
Not having luck with manual testing. The video module is active too, maybe Jetpack Beta is pulling a stale copy. Did this work for you @dmsnell ?

<img width="956" alt="screen shot 2019-01-24 at 2 00 02 pm" src="https://user-images.githubusercontent.com/1270189/51711245-5cb4d100-1fe0-11e9-9209-25339b9bb5b4.png">

jeherve commented on modules/videopress/class.videopress-gutenberg.php:
```diff
@@ -0,0 +1,84 @@
+	 * Register the Jetpack Gutenberg extension that adds VideoPress support to the core video block.
+	 */
+	public static function register_video_block_with_videopress() {
+		register_block_type( 'core/video', array(
```
This may be a stupid question as I did not follow all the conversations about all this, but could we just rely on `Jetpack_Gutenberg` to do all the heavy lifting here?

Instead of using `register_block_type`, could we use `jetpack_register_block`, and that would thus cover all the things we check in `should_initialize` above? See how it's done for the Contact Form block for example:
https://github.com/Automattic/jetpack/blob/60026de776b505d42add576a9eee00a108bd1b5d/modules/contact-form/grunion-contact-form.php#L239

gwwar replied:
See pafL3P-gg-p2 p7jreA-264-p2, the TL;DR is that these VideoPress cases fall outside of the current Jetpack registration assumptions. This patch is intended to be a quicker fix for the user, while we resolve conversations of what we need next. (Either more core contribs or framework investments).

dmsnell replied:
we should add a comment here too to warn future devs from coming back around and trying to "fix" this or update to the `jetpack_register_block()` so that they don't start down that path and waste time before realizing it won't work

mmtr replied:
Comment added in 2d9d076

jeherve commented on modules/videopress/class.videopress-gutenberg.php:
```diff
@@ -0,0 +1,84 @@
+			'/src="([^"]+)/',
+			sprintf(
+				'src="%1$s',
+				esc_attr( $videopress_url )
```
Should we use `esc_url` here instead?

dmsnell replied:
good call! though as a `src` it looks like the WordPress docs suggest that `esc_url_raw()` might be more fitting, the difference being the replacement of entities for display or not

`esc_url_raw()`
> The resulting URL is safe to use in database queries, redirects and HTTP requests.

jeherve commented on modules/videopress/class.videopress-gutenberg.php:
```diff
@@ -0,0 +1,84 @@
+		$videopress_id = video_get_info_by_blogpostid( $blog_id, $post_id )->guid;
+		$videopress_data = videopress_get_video_details( $videopress_id );
+
+		if ( empty( $videopress_data->files->hd->mp4 ) ) {
```
Should we check for `$videopress_data->file_url_base->https` as well, since we also use it later?

jeherve commented on modules/videopress/class.videopress-gutenberg.php:
```diff
@@ -0,0 +1,84 @@
+			return $content;
+		}
+
+		$blog_id = get_current_blog_id();
```
You'll probably need to align all those assignments to make our linter happy.

jeherve reviewed (changes requested):
I have a few questions about the implementation here.

dmsnell commented on modules/videopress/class.videopress-gutenberg.php:
```diff
@@ -0,0 +1,84 @@
+			$content,
+			1
+		);
+	}
```
While here I have a point I thought we might address, which is to add the original movie source. I noticed that we do that with images that get replaced by WordPress and it seems like a valid thing to do here. On one hand I don't want to delay this work but on the other hand it seems like it would be a convenient place to make the change.

We'd have to _either_ update the RegEx pattern or switch to real HTML parsing, which honestly would have other benefits but obviously would delay the work.

```php
preg_replace(
	'/(\s)(?:src=(['"])(?:(?!\2.)+?)\2(\s)/',
	sprintf(
		'\1src="%1$s" data-orig-file="%2$s"\3',
		esc_url_raw( $videopress_url ),
		esc_url_raw( $original_url ),
	)
);
```

mmtr replied:
Sounds like a good enhancement! However I'm having hard times try to get the original url from the regular expression. I think we should get it from a matched group, but it seems that a group starting with `?:` is a non capturing group, so I don't know how can we define a 3rd group that captures the original src attribute value.

dmsnell replied:
> a group starting with ?: is a non capturing group

that's correct. we can make it a capture group by removing the `?:` which tells it to ignore the group.

`(\s)(?:src=(['"]))((?:(?!\2).)+?)\2([\s\/>])`

note that I think I made a mistake in the original one there - [see it live at Regex101](https://regex101.com/r/CMSI64/1)

once we have the original too we could always determine if it needs replacement and early-abort again. if we do this before requesting the metadata we can abort before running those more expensive queries. this would address the case where the media inserter adds the VideoPress URLs already…

```php
function replace_src_preg_callback( $matches ) {
	list( $full_match, $prefix_space, $quote_char, $src, $suffix ) = $matches;

	if ( 0 === strstr( $src, 'https://videos.files.wordpress.com/' ) ) {
		return $full_match;
	}

	$blog_id = …
	$post_id = …

	$videopress_data = …

	if ( empty( $videopress_data->… ) ) {
		return $full_match;
	}

	return $prefix_space . 'src="' . $new_src . '"' . $suffix;
}

…

preg_replace_callback(
	self::VIDEO_SOURCE_PATTERN,
	array( 'self', 'replace_src_preg_callback' ),
	$content,
	/* limit */ 1
); 
```

dmsnell replied:
noting how complicated this RegExp pattern is getting even while matching a very simple thing is why I like to encourage us to explore methods that are reliable and more tuned to actual HTML parsing, even though they might sound scary - with small input documents I suspect they bring no performance issues, plug we can preserve the existing markup and not worry about oddball cases such as when we encounter `<video bad-src="bad.mp4" src="good.mp4" />`

gwwar replied:
Ah, I asked a similar question in D23482-code, and it looks like this is quickly increasing in complexity. 

Provided that the related HTML parser isn't prohibitively slow, I think it's safer to use one rather than carry on with the regex matching. Within this function we have a reasonable guarantee that we only are passed the block html?

dmsnell replied:
we have a full guarantee to only get the block contents. note too that we could also filter the block were we to want to, but `render_callback` is also sufficient. the only time we'd expect to get lots of HTML is if someone manually makes the caption enormous.

dmsnell replied:
not sure cc: @dereksmart @sirreal - most of what I've seen is RegExp work but also most of what I've seen is at risk for these same kind of risks of complex cases defeating the RegExp _or_ of making the RegExp pattern hard to read in an attempt at catching those complex cases

gwwar replied:
Since this is dragging out a bit, let's go back to this as a follow up. By not embedding the original src url, is this something that's lost forever or can we look that up through alternative means?

dmsnell replied:
we can avoid the concerns for now and stick to basic replacement

mmtr replied:
> By not embedding the original src url, is this something that's lost forever or can we look that up through alternative means?

It'll be still on the block content stored on the post content.

That's why I think we should not worry about the concerns because we're going to change the post content anyway in future iterations when the block content will render the VideoPress player instead of the native `video` HTML5 player.

dmsnell commented on modules/videopress/class.videopress-gutenberg.php:
```diff
@@ -0,0 +1,84 @@
+		$videopress_url = $videopress_data->file_url_base->https . $videopress_data->files->hd->mp4;
+
+		return preg_replace(
+			'/src="([^"]+)/',
```
we probably need to update this pattern. we don't have to be fully-spec compliant but we at least need to make sure we don't accidentally replace the wrong attributes, such as `<video alt-src="…">` or `orig-src="…"`

```php
// fixing the bug at hand - make sure whitespace precedes the attribute
$pattern = '/\ssrc="([^"]+)/';

// allowing for valid single quotes - note the lazy expansion
$pattern = '/\ssrc=(['"])(?:(?!\1).)+?\1/';

// preserve the initial space when we rewrite
$pattern = '/(\s)src=(['"])(?:(?!\2).)+?\2/'

preg_replace( $pattern, sprintf( '\1src="%1$s"', esc_url_raw( $videopress_url ) ) );
```

in this last iteration we're preserving as much of the original as we can and more resilient against failing to replace the src just because someone used single-quotes

dmsnell commented:
Status update on my testing: I spent a lot of time testing this on Jurassic Ninja just to discover that none of the renders were working as I expected, so I installed a local copy of Jetpack and started doing some diagnosis to find an error when retrieving the VideoPress data…

```
$videopress_id = {
	"post_id": 11,
	"blog_id": 1,
	"guid": null,
	"finish_date_gmt": "0000-00-00 00:00:00"
}->guid;

$videopress_data = {
	"errors": {
	"bad-guid-format": [
			"Invalid Video GUID!"
		]
	},
	"error_data": []
}
```

The `blog_id` is `1` so I'm thinking that maybe we're just asking Jetpack for its own site id instead of asking for the WordPress.com site id.

When I replaced `get_current_blog_id()` with `Jetpack_Options::get_option( 'id' )` I got the expected blog/site id that I wanted. Sadly I still had a `NULL` `guid` but I was also having trouble uploading my video files - maybe due to my internet, maybe due to `ngrok`

dmsnell committed:
Try and get proper WordPress.com site id

This commit may be wrong but I'm learning how to properly get the
expected site id and this seemed to get it. I came to it from looking at
some other code - copy/paste.

`get_current_site_id()` was returning `1` even though my actual
WordPress.com site id was something expected, a very large number.

dmsnell commented:
@mmtr I added a commit here to try and test my changes. I still am having trouble uploading a video, getting an HTTP error on every attempt though it did work before. I'm stumped.

mmtr committed:
Check that the https file URL base is not empty

mmtr committed:
Clarify why we don't use `jetpack_register_block`

mmtr committed:
Handle src attributes with single quotes

mmtr committed:
Use `get_current_blog_id` in WP.com environments

mmtr committed:
Fix alignment in assignments

dmsnell commented:
My HTML parsing example seemed to have gotten lost, so I'm including here for posterity's sake.

```php
$content = '<!-- wp:video -->
<figure class="wp-block-video"><video controls src="blob:https://responsible-mongoose-462.jurassic.ninja/977efa53-e85a-4693-8fbe-287ff006b5ea"></video></figure>
<!-- /wp:video -->';

function null_error_handler() {
  return true;
}

function modify_html_fragment( $f, $html ) {
  $doc = new DOMDocument();
  $doc->preserveWhiteSpace = true;
  $doc->formatOutput = true;

  // hide warnings about invalid HTML <video> tag
  $error_handler = set_error_handler( 'null_error_handler' );
  $doc->loadHTML( '<!DOCTYPE html5><div>' . $html . '</div>' );
  set_error_handler( $error_handler );

  $f( $doc );

  $div = $doc->getElementsByTagName( 'div' )[0];
  $output = '';
  foreach ( $div->childNodes as $child ) {
    $output .= $doc->saveHTML( $child );
  }
  return $output;
}

return modify_html_fragment( function( $doc ) {
  $video = $doc->getElementsByTagName( 'video' )[ 0 ];
  $old_src = $video->getAttribute( 'src' );
  $video->removeAttribute( 'src' );

  $source = $doc->createElement( 'source' );
  $source->setAttribute( 'src', $old_src );
  $video->appendChild( $source );
}, $content );

/*
<!-- wp:video -->
<figure class="wp-block-video"><video controls><source src="blob:https://responsible-mongoose-462.jurassic.ninja/977efa53-e85a-4693-8fbe-287ff006b5ea"></source></video></figure><!-- /wp:video -->
*/
```

obviously the anonymous function needs to go away to make it work, but I separated out the process of modifying HTML from the logic of what we're modifying. from inside of `modify_html_fragment()` we can do whatever we want with the attributes and children.

mmtr committed:
Remove unnecessary checks

mmtr commented:
@dmsnell @gwwar I think we shouldn't worry about storing the original `src` attribute. This is just a temporary patch and in future iterations we're going to replace the whole `<video>` markup with the VideoPress player (hopefully reusing the same code we use for the `wpvideo` shortcode).

dmsnell commented:
Here's a test post I used to confirm behaviors.

```html
<!-- wp:paragraph -->
<p>Uploaded before the plugin was activated.</p>
<!-- /wp:paragraph -->

<!-- wp:video {"id":6} -->
<figure class="wp-block-video"><video controls src="http://17588768.ngrok.io/wp-content/uploads/2019/01/SampleVideo_1280x720_5mb.mp4"></video></figure>
<!-- /wp:video -->

<!-- wp:paragraph -->
<p>Uploaded with plugin, Media Library > Upload</p>
<!-- /wp:paragraph -->

<!-- wp:video {"id":10} -->
<figure class="wp-block-video"><video controls src="https://17588768.ngrok.io/samplevideo_1280x720_5mb-mp4/"></video></figure>
<!-- /wp:video -->

<!-- wp:paragraph -->
<p>Same attachment, added from Media Library after transcription</p>
<!-- /wp:paragraph -->

<!-- wp:video {"id":10} -->
<figure class="wp-block-video"><video controls src="https://videos.files.wordpress.com/03H0DvlD/samplevideo_1280x720_5mb_hd.mp4"></video></figure>
<!-- /wp:video -->

<!-- wp:paragraph -->
<p>Uploaded with <em>Upload</em></p>
<!-- /wp:paragraph -->

<!-- wp:video {"id":14} -->
<figure class="wp-block-video"><video controls src="http://17588768.ngrok.io/wp-content/uploads/2019/01/SampleVideo_1280x720_5mb-2.mp4"></video></figure>
<!-- /wp:video -->

<!-- wp:paragraph -->
<p>Same attachment added after <em>Upload</em></p>
<!-- /wp:paragraph -->

<!-- wp:video {"id":16} -->
<figure class="wp-block-video"><video controls src="http://17588768.ngrok.io/wp-content/uploads/2019/01/SampleVideo_1280x720_5mb-2.mp4"></video></figure>
<!-- /wp:video -->
```

In `jetpack/master` these were the URLs in order that were produced. The VideoPress URL is there because the media library added it.
```
http://17588768.ngrok.io/wp-content/uploads/2019/01/SampleVideo_1280x720_5mb.mp4
https://17588768.ngrok.io/samplevideo_1280x720_5mb-mp4/
https://videos.files.wordpress.com/03H0DvlD/samplevideo_1280x720_5mb_hd.mp4
http://17588768.ngrok.io/wp-content/uploads/2019/01/SampleVideo_1280x720_5mb-2.mp4
http://17588768.ngrok.io/wp-content/uploads/2019/01/SampleVideo_1280x720_5mb-2.mp4
```

In this branch we can see one more video gets transformed with the VideoPress URL as we expect.
```
http://17588768.ngrok.io/wp-content/uploads/2019/01/SampleVideo_1280x720_5mb.mp4
https://videos.files.wordpress.com/03H0DvlD/samplevideo_1280x720_5mb_hd.mp4
https://videos.files.wordpress.com/03H0DvlD/samplevideo_1280x720_5mb_hd.mp4
http://17588768.ngrok.io/wp-content/uploads/2019/01/SampleVideo_1280x720_5mb-2.mp4
http://17588768.ngrok.io/wp-content/uploads/2019/01/SampleVideo_1280x720_5mb-2.mp4
```

The Core video block can run into issues with the way it stores the video `src` as a `blob:` URL until the video is uploaded. If it never finished uploading then we can end up leaving the editor in a state that references that possibly-invalid in-memory data URL. Of course, it will be invalid for every site visitor in any case.

This confirms that we are only rendering videos which were uploaded through *Media Library* > *Upload* and _after_ the Jetpack plugin and HD video setting were enabled. Even if it's enabled when inserting the video into the post, it must also be uploaded or reuploaded to make it into VideoPress.

mmtr commented:
@dmsnell I wouldn't consider that a blocker for this PR, mainly because that misleading behavior is already happening in master and it is not related to this PR.

The changes proposed on this PR don't change the way the videos are uploaded to VideoPress. It just replaces the `src` attribute with the VideoPress URL if the video has been uploaded to VideoPress.

How the videos are uploaded to VideoPress is handled by a different logic and we certainly have issues there:
- Videos uploaded in "Media > Add" are not uploaded to VideoPress: https://github.com/Automattic/jetpack/issues/7073
- Videos uploaded using the "Upload" button from the core video block are not uploaded to VideoPress: https://github.com/Automattic/jetpack/issues/11194

This means that right now, only the videos uploaded in "Media > Library" or the ones using the "Media Library" button from the core video block are uploaded to VideoPress.

dmsnell commented:
No need to block on existing behaviors @mmtr - I was only trying to be thorough to document the behaviors. The last item may have gone unnoticed though due to my poor wording.

> Once I removed the plan from my test account but left Jetpack and the feature enabled then I was still getting the VideoPress replacement URLs. This should be disabled when the plan disappears.

 - I added a video through the expected way and it uploaded to VideoPress.
 - The URL in the post was the site-local address `https://17588768.ngrok.io/samplevideo_1280x720_5mb-mp4/`
 - As expected, this was replaced on render with the VideoPress URL
 - Once I removed the plan from my account though it was still replacing the URL on render

This could be caused by the related work going on with extension availability. If that's the case ignore this one too. If it's not the case then we probably want to make sure we don't keep the feature going when the necessary plan is gone. If this isn't related to the availability work then we could also be facing a cache issue.

---

Actually when reviewing the code I think this is all related to the availability work and so probably we just need to let people know that it's not dependent on the plan yet.

gwwar reviewed (approved):
Thanks @mmtr this is a very good first step ✨ 

I did a second round of manual testing and verified that without the plan setup, uploaded videos are stored on the site, and after a plan _future_ uploads are sent to VideoPress. (Existing ones are left alone).

<img width="338" alt="screen shot 2019-01-30 at 1 51 14 pm" src="https://user-images.githubusercontent.com/1270189/52015464-e3612680-2496-11e9-9586-e4473d726845.png">

Working video player, on unsupported browser file type.

<img width="968" alt="screen shot 2019-01-30 at 1 51 52 pm" src="https://user-images.githubusercontent.com/1270189/52015496-fa077d80-2496-11e9-8f94-1ec46dca83f0.png">

gwwar commented:
For any odd existing behaviors, let's be sure to create issues or add details to existing ones so we can properly follow up on that.

gwwar commented:
@jeherve this is ready for another look when you have a chance.

mmtr commented:
> Once I removed the plan from my test account but left Jetpack and the feature enabled then I was still getting the VideoPress replacement URLs. This should be disabled when the plan disappears.

Oh right, I can reproduce this. It should be disabled indeed, but I'm starting to think that this is unrelated to this PR.

Mainly, because the replacement is included in the `/modules/videopress/class.videopress-gutenberg.php` file, loaded by the `/modules/videopress.php` file, which I expect to don't be loaded on free plans.

So, the issue is that Jetpack loads the `/modules/videopress.php` file after removing the Premium plan from the account. Trying to find what is causing this, I noted that as soon as you visit "Jetpack > Settings" the replacement is no longer happening, so it seems Jetpack caches somewhere that the VideoPress module is active and doesn't inactivate it until we visit "Jetpack > Settings". 

Will investigate this a bit further and file a separate issue.

jeherve commented:
> Repeat the process downgrading the plan or disabling the VideoPress module and check that the video is not playable on the published view and the src attribute doesn't point to videos.files.wp.com. Note that you need to upload a new video since the previous one is already converted).

I'd like to come back to this. I think it is somewhat related to the comment I left here:
https://github.com/Automattic/jetpack/pull/11162#pullrequestreview-198553050

The way VideoPress works today (on wpcom and in Jetpack), once a video has been uploaded to VideoPress it's there forever. The video is not deleted from our servers when your plan expires. This also allows others to embed the videos you've uploaded on their own sites, via the share options:

![image](https://user-images.githubusercontent.com/426388/52053689-5aabbe80-255a-11e9-9dc8-0c36b196d7b0.png)

This is why, even if the VideoPress module is deactivated from your site because your plan expired, we still allow you to embed videos on your site, either thanks to [the VideoPress oEmbed in Core](https://github.com/WordPress/WordPress/blob/60ddb6012e14fb2791edcdc29358397aa0c0e151/wp-includes/class-oembed.php#L90), or via the VideoPress shortcode. That shortcode is included when the VideoPress module is active:
https://github.com/Automattic/jetpack/blob/f098dbe15b348e8d41a1ef97ea65b4e8d7224900/modules/videopress.php#L16
But it is also included if the VideoPress module is not active anymore: 
https://github.com/Automattic/jetpack/blob/87cc282f9157dfd81260ac8ae048c7c830f4d983/modules/shortcodes/videopress.php#L18

This way, **we never break video embeds on a site, even if you stop paying.** I say never, but in fact if you use the shortcode and then later on deactivate Jetpack altogether (or even just deactivate the Shortcode embeds module) you will lose your videos. This is why I was previously pushing us to move away from the shortcodes, and towards using full URLs that can be processed as oEmbeds: #6587

I think we should also aim to not break videos on a site, even after you stop paying. 

In practice, maybe that would mean including `videopress/class.videopress-gutenberg.php` in `modules/shortcodes/videopress.php` alongside the shortcode and `modules/videopress/utility-functions.php`. Maybe that should even be done in `modules/blocks.php`, so this all happens even when the Shortcodes module is not active.

The problem with this approach is that we can't assume that the site is connected anymore; `Jetpack_Options::get_option( 'id' )` may not return any blog ID. I think that brings us back to what I was mentioning in https://github.com/Automattic/jetpack/pull/11162#pullrequestreview-198553050

What do you all think about this?

jeherve commented:
That makes sense, thank you. In this case, I think we're good to go! Merge when you're ready!

jeherve reviewed (approved):
Tests well now. 👍 	🚢
